### ISSUE TITLE
Constantly constant suggestions

### ISSUE THREAD

**Issue by hnry** [joefitzgerald/autocomplete-go#2]
It's stuck trying to suggest me things even though it's not obvious to me why?
It happens on default settings `autocomplete-plus`. Tried playing around with settings with no luck (except disable suggestions entirely). 

Example 1: In comments, its constantly suggesting.
![wtf2](https://cloud.githubusercontent.com/assets/25837/12740165/a36e1724-c93e-11e5-82e2-d44e20e99072.gif)

Example 2: In code, it's constantly suggesting on blank spaces or non alphabetic characters that should have no match.
![wtf](https://cloud.githubusercontent.com/assets/25837/12740228/2a478780-c93f-11e5-91af-cb9809f6b2df.gif)

Besides it taken screen space for no reason, it makes tabbing and pressing enter a nightmare.

As an unrelated side note, I don't think it should be suggesting Test\* functions

**Comment by joefitzgerald** [comment]
Sorry you're having an issue @hnry! May I ask - do you also have go-plus installed? Are both autocomplete-go and go-plus enabled?

This package will eventually replace the autocomplete+ provider built-in to go-plus but requires a little more maturation before it's ready for that.

Could you provide more information? 
- What version of go are you using? 
- When did you last run `gocode close && go get -u github.com/nsf/gocode`? 
- Do you also use Sublime (which has its own embedded version of gocode that can cause a conflict)?
- Are you using go to build, or gb?
- Are you using GO15VENDOREXPERIMENT?
- Any other useful context or steps to reproduce from a new file would be helpful.

Thanks for the feedback and again, sorry for the inconvenience!

**Comment by hnry** [comment]
@joefitzgerald  thanks for the quick response, I do not have go-plus installed. only autocomplete-go and go-config (which i believe was installed by autocomplete-go).
- go version go1.5.3 darwin/amd64  (OS X 10.11.3)
- I installed gocode about 1-2 weeks ago, but never ran a update
- I don't have Sublime installed
- unsure so probably no
- unsure so probably no

I did a fresh install of OS X about 2-3 weeks ago, and decided to also give atom a try. So most things are pretty up to date.

As for atom itself, I installed a few packages. But for the most part everything is on default. 
I removed gocode and reinstalled with go get.
I tried it with disabling other autocomplete packages for other languages.
I tried uninstalling autocomplete-go and go-config through atom package manager, restarted atom, then reinstalled, then restarted to test, same thing.

Even with all project folders removed, all windows closed, just a single barebones go file opened it does it then too for me. 

_update_: I disabled all community packages, only leaving autocomplete-go, go-config, and environment. still same thing.

_update_: I'm running gocode in debugging mode and it shows the "bad" suggestions there, but that doesn't give me any hints.

**Comment by hnry** [comment]
I am surprised @joefitzgerald that you don't experience the same issue on your end. Because most of my setup is new since I just did a fresh install 2-3 weeks ago. But just to be _extra_ sure, I installed a windows 7 virtual machine and installed the bare minimum to use this plugin and I am seeing the same thing there.

![wtf3](https://cloud.githubusercontent.com/assets/25837/12769303/bc02bc26-c9e4-11e5-9c67-bd268b3d28df.gif)

_So this issue should be experienced by everyone_.

For GoSublime plugin for Sublime I do not have this issue.
gocode -debug -s seems to be reporting the same results as far as giving _everything_ as candidates if the cursor is not even on anything relevant (like white space). I think this is normal and not a bug with gocode.

So I _think_ GoSublime just has a better way of not asking gocode for candidates for every possible cursor movement or has a better way of interpreting the results if it does always ask for candidates.
- if the keypress is a blank space or backspace or invalid symbol it should not ask gocode for candidates
- or use gocode results better to determine if it's an error and if gocode is just suggesting the entire context of the file as candidates
- it should respect autocomplete's minimum length for completion setting (unless it's a go built in or imported std library?)

**Comment by joefitzgerald** [comment]
So you're in Windows. And I'm on OSX. That could explain it. I'll try and reproduce tonight or early tomorrow on a VM.

**Comment by hnry** [comment]
No, I am on OS X too. (The first 2 gifs are on OS X)
The 3rd gif I posted that's on Windows is me setting up a brand new Windows 7 VM to reproduce. I setup a Windows 7 VM for the sole purpose of seeing if a completely fresh install would do the same thing. It does. It has all default configs and only installed the minimum required to get autocomplete-go to run.

So on both setups, OS X 10.11.3 (clean install that's 2-3 weeks old), or Windows 7 (freshly installed VM) I am seeing the same thing.

BTW go-plus does not do the same thing for me on my computer. (Neither does GoSublime)

**Comment by joefitzgerald** [comment]
Reproduced. Will try to fix this tomorrow.

**Comment by joefitzgerald** [comment]
I can reproduce this in go-plus also. This appears to be a gocode thing. I typed the exact same set of characters into the editor as you did in:

![](https://cloud.githubusercontent.com/assets/25837/12769303/bc02bc26-c9e4-11e5-9c67-bd268b3d28df.gif)

I did this with:
- `go-plus` disabled and `autocomplete-go` enabled
- `go-plus` enabled and `autocomplete-go` disabled

Same result in both cases.

**Comment by joefitzgerald** [comment]
@hnry I've made the package configurable, using similar options to those I provided in `go-plus`. This should allow you to suppress scopes (e.g. comment is suppressed by default). To find the current scope for the cursor, type CMD-Option-P. 

The changes in `0.2.0` also include exclusion of autocomplete on some characters (including `space`, `comma`, `newline`, et al. Take a look at the updated readme or the package settings.
